lib/rules/semi-style.js

```diff
@@ -0,0 +1,131 @@
+/**
+ * @fileoverview Rule to enforce location of semicolons.
+ * @author Toru Nagashima
+ */
+
+"use strict";
+
+//------------------------------------------------------------------------------
+// Requirements
+//------------------------------------------------------------------------------
+
+const astUtils = require("../ast-utils");
+
+//------------------------------------------------------------------------------
+// Rule Definition
+//------------------------------------------------------------------------------
+
+const SELECTOR = `:matches(${
+    [
+        "BreakStatement", "ContinueStatement", "DebuggerStatement",
+        "EmptyStatement", "ExportAllDeclaration", "ExportDefaultDeclaration",
+        "ExportNamedDeclaration", "ExpressionStatement", "ImportDeclaration",
+        "ReturnStatement", "ThrowStatement", "VariableDeclaration"
+    ].join(",")
+})`;
+
+module.exports = {
+    meta: {
+        docs: {
+            description: "enforce location of semicolons",
+            category: "Stylistic Issues",
+            recommended: false
+        },
+        schema: [{
+            anyOf: [
+                { enum: ["last", "first"] },
+                {
+                    type: "object",
+                    properties: {
+                        statements: { enum: ["last", "first"] },
+                        forLoopHead: { enum: ["last", "first"] }
+                    },
+                    additionalProperty: false
+                }
+            ]
+        }],
+        fixable: "whitespace"
+    },
+
+    create(context) {
+        const sourceCode = context.getSourceCode();
+        const options = context.options[0] || "last";
+        const statementLoc = typeof options === "string" ? options : (options.statements || "last");
+        const forLoopHeadLoc = typeof options === "string" ? options : (options.forLoopHead || "last");
+
+        /**
+         * Check whether comments exist between the given 2 tokens.
+         * @param {Token} left The left token to check.
+         * @param {Token} right The right token to check.
+         * @returns {boolean} `true` if comments exist between the given 2 tokens.
+         */
+        function commentsExistBetween(left, right) {
```

Comment:
This function gets created every time rule context is created. Why not just move this function outside so that single declaration can used across different context as this function is pure (or can be made pure).
I know we do this in almost all rules but I think we should change this style.
thoughts?